nil/services/nilservice/service.go

```diff
@@ -461,34 +486,44 @@ func createShards(
 	collatorTickPeriod := time.Millisecond * time.Duration(cfg.CollatorTickPeriodMs)
 	syncerTimeout := syncTimeoutFactor * collatorTickPeriod
 
-	funcs := make([]concurrent.Func, 0, 2*cfg.NShards)
+	funcs := make([]concurrent.Func, 0, 2*cfg.NShards+1)
+	syncers := make([]*collate.Syncer, 0, cfg.NShards)
 	pools := make(map[types.ShardId]txnpool.Pool)
 
-	var wgFetch sync.WaitGroup
-	wgFetch.Add(int(cfg.NShards))
+	var wgInit sync.WaitGroup
+	wgInit.Add(1)
 
 	pKey, err := cfg.LoadValidatorPrivateKey()
 	if err != nil {
 		return nil, nil, err
 	}
 
-	if !cfg.SplitShards && len(cfg.Validators) == 0 {
-		pubkey, err := cfg.ValidatorKeysManager.GetPublicKey()
-		if err != nil {
-			return nil, nil, err
-		}
-		for i := range cfg.NShards {
-			cfg.Validators[types.ShardId(i)] = []config.ValidatorInfo{
-				{PublicKey: config.Pubkey(pubkey)},
+	if cfg.ZeroState == nil {
+		cfg.ZeroState = &execution.ZeroStateConfig{}
+	}
+
+	if !cfg.SplitShards {
+		if len(cfg.ZeroState.ConfigParams.Validators.Validators) == 0 {
```

Comment:
Let's implement a getter for that

```suggestion
		if len(cfg.ZeroState.GetValidators()) == 0 {
```